### PR TITLE
Architecture: Flatten Heap Substates

### PR DESCRIPTION
**Pull request by talekhinezh**
* Flatten Heap to substates so that it's structure is closer to the Store
* Refactored DropNode logic
* Add DropNodeVisibility property

### CODE REVIEW

File: radix-engine/src/engine/heap.rs [hunk 1]
```diff
@@ -112,3 +149,63 @@ impl Heap {
...
+}
+
+impl Into<ProofSubstate> for HeapRENode {
+    fn into(mut self) -> ProofSubstate {
+        self.substates
+            .remove(&SubstateOffset::Proof(ProofOffset::Proof))
+            .unwrap()
+            .into()
+    }
+}
```
Comment (iamyulong) [inline radix-engine/src/engine/heap.rs:211]: This is a general comment. I think we should use `TryInto` trait for conversions that may fail.

File: radix-engine/src/engine/kernel.rs [hunk 2]
```diff
@@ -339,6 +339,23 @@ where
...
+        for node_id in owned_nodes {
+            if let RENodeId::Worktop = node_id {
+                worktops.push(node_id);
+            } else {
+                self.drop_node(node_id)?;
+            }
+        }
+        for worktop_id in worktops {
+            self.drop_node(worktop_id)?;
+        }
```
Comment (iamyulong) [inline radix-engine/src/engine/kernel.rs:354]: This is interesting. Why do we need to drop other nodes first? For the proofs?

File: radix-engine/src/engine/kernel.rs [hunk 3]
```diff
@@ -949,7 +970,10 @@ where
 
         // TODO: Move this into higher layer, e.g. transaction processor
         if self.current_frame.depth == 0 {
-            self.current_frame.drop_frame(&mut self.heap)?;
+            self.current_frame.drop_all_locks(&mut self.track)?;
+            self.execution_mode = ExecutionMode::Application;
+            self.drop_nodes_in_frame()?;
+            self.execution_mode = ExecutionMode::Kernel;
```
Comment (iamyulong) [inline radix-engine/src/engine/kernel.rs:976]: I think the current of execution mode management is very fragile and error prone. Needs a better machanism for execution mode switching. :)

File: radix-engine/src/model/substates/package.rs [hunk 4]
```diff
@@ -1,14 +1,21 @@
 use sbor::rust::fmt::Debug;
+use std::fmt::Formatter;
```
Comment (iamyulong) [inline radix-engine/src/model/substates/package.rs:2]: Minor: can we use `sbor::rust::fmt::Formatter;`?

**Review by iamyulong — COMMENTED** [review]
LGTM with minor comments.

File: radix-engine/src/engine/kernel.rs (hunk 3 above)
Comment (talekhinezh) [reply radix-engine/src/engine/kernel.rs:976]: Yup, I didn't totally finish this part, just wanted to get the PR out as soon as possible :) will cleanup :)

**Review by iamyulong — APPROVED** [review]

**Pull request merged**